## FLIGHT PREDICTION DATA SET EDA AND FEATURE ENGINEERING 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
# HERE WE HAVE TESTING AND TRAINING DATA GIVEN TO US 
# SO IMPORTING THE  DATA SETS

In [3]:
train_data = pd.read_excel('Data_Train.xlsx')
train_data.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [4]:
test_data = pd.read_excel('Test_set.xlsx')
test_data.head()

Airline Date_of_Journey    Source Destination            Route  \
0        Jet Airways       6/06/2019     Delhi      Cochin  DEL → BOM → COK   
1             IndiGo      12/05/2019   Kolkata    Banglore  CCU → MAA → BLR   
2        Jet Airways      21/05/2019     Delhi      Cochin  DEL → BOM → COK   
3  Multiple carriers      21/05/2019     Delhi      Cochin  DEL → BOM → COK   
4           Air Asia      24/06/2019  Banglore       Delhi        BLR → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops              Additional_Info  
0    17:30  04:25 07 Jun  10h 55m      1 stop                      No info  
1    06:20         10:20       4h      1 stop                      No info  
2    19:15  19:00 22 May  23h 45m      1 stop  In-flight meal not included  
3    08:00         21:00      13h      1 stop                      No info  
4    23:55  02:45 25 Jun   2h 50m    non-stop                      No info

In [5]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [6]:
train_data.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

## CHECKING FOR NULL VALUES 
Here we have only one null vlaue in both route and total stops 
as the data set is having total of 10683 entries , 
we can drop one entry from the data set to eliminate the null values

In [7]:
train_data.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

In [8]:
train_data.shape

(10683, 11)

In [9]:
train_data[train_data['Route'].isnull() == True ]       # NULL Value present in the below given entry

Airline Date_of_Journey Source Destination Route Dep_Time  \
9039  Air India       6/05/2019  Delhi      Cochin   NaN    09:45   

      Arrival_Time Duration Total_Stops Additional_Info  Price  
9039  09:25 07 May  23h 40m         NaN         No info   7480

In [10]:
train_data.dropna(axis=0, inplace=True)

In [11]:
train_data.shape

(10682, 11)

In [12]:
train_data.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              0
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        0
Additional_Info    0
Price              0
dtype: int64

## COMBINING THE TRAINING AND TESTING DATA TO PERFORM EDA

In [13]:
data = train_data._append(test_data)

In [14]:
data.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info    Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897.0  
1    05:50         13:15   7h 25m     2 stops         No info   7662.0  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882.0  
3    18:05         23:30   5h 25m      1 stop         No info   6218.0  
4    16:50         21:35   4h 45m      1 stop         No info  13302.0

In [15]:
data.tail()

Airline Date_of_Journey   Source Destination            Route  \
2666          Air India       6/06/2019  Kolkata    Banglore  CCU → DEL → BLR   
2667             IndiGo      27/03/2019  Kolkata    Banglore        CCU → BLR   
2668        Jet Airways       6/03/2019    Delhi      Cochin  DEL → BOM → COK   
2669          Air India       6/03/2019    Delhi      Cochin  DEL → BOM → COK   
2670  Multiple carriers      15/06/2019    Delhi      Cochin  DEL → BOM → COK   

     Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
2666    20:30  20:25 07 Jun  23h 55m      1 stop         No info    NaN  
2667    14:20         16:55   2h 35m    non-stop         No info    NaN  
2668    21:50  04:25 07 Mar   6h 35m      1 stop         No info    NaN  
2669    04:00         19:15  15h 15m      1 stop         No info    NaN  
2670    04:55         19:15  14h 20m      1 stop         No info    NaN

## EXPLORING DIFFERENT FEATURES OF THE DATA SET

In [16]:
data.Source.value_counts()

Source
Delhi       5681
Kolkata     3581
Banglore    2752
Mumbai       883
Chennai      456
Name: count, dtype: int64

In [17]:
data.Destination.value_counts()

Destination
Cochin       5681
Banglore     3581
Delhi        1582
New Delhi    1170
Hyderabad     883
Kolkata       456
Name: count, dtype: int64

In [18]:
data.Total_Stops.value_counts()

Total_Stops
1 stop      7056
non-stop    4340
2 stops     1899
3 stops       56
4 stops        2
Name: count, dtype: int64

In [19]:
data.Airline.value_counts()

Airline
Jet Airways                          4746
IndiGo                               2564
Air India                            2191
Multiple carriers                    1543
SpiceJet                             1026
Vistara                               608
Air Asia                              405
GoAir                                 240
Multiple carriers Premium economy      16
Jet Airways Business                    8
Vistara Premium economy                 5
Trujet                                  1
Name: count, dtype: int64

In [20]:
data.Additional_Info.value_counts()

Additional_Info
No info                         10492
In-flight meal not included      2426
No check-in baggage included      396
1 Long layover                     20
Change airports                     8
Business class                      5
No Info                             3
1 Short layover                     1
Red-eye flight                      1
2 Long layover                      1
Name: count, dtype: int64

## SPLITTING DATE OF JOURNEY INTO DAY MONTH AND YEAR IN INTEGER FORMAT

In [21]:
day = []
month = []
year = []

for date in data['Date_of_Journey']:
    date = str(date).split('/')
    day.append(int(date[0]))
    month.append(int(date[1]))
    year.append(int(date[2]))

data.insert(loc=1, column='JR_Day', value=day)
data.insert(loc=2, column='JR_Month', value=month)
data.insert(loc=3, column='JR_Year', value=year)

In [22]:
data.head()

Airline  JR_Day  JR_Month  JR_Year Date_of_Journey    Source  \
0       IndiGo      24         3     2019      24/03/2019  Banglore   
1    Air India       1         5     2019       1/05/2019   Kolkata   
2  Jet Airways       9         6     2019       9/06/2019     Delhi   
3       IndiGo      12         5     2019      12/05/2019   Kolkata   
4       IndiGo       1         3     2019      01/03/2019  Banglore   

  Destination                  Route Dep_Time  Arrival_Time Duration  \
0   New Delhi              BLR → DEL    22:20  01:10 22 Mar   2h 50m   
1    Banglore  CCU → IXR → BBI → BLR    05:50         13:15   7h 25m   
2      Cochin  DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h   
3    Banglore        CCU → NAG → BLR    18:05         23:30   5h 25m   
4   New Delhi        BLR → NAG → DEL    16:50         21:35   4h 45m   

  Total_Stops Additional_Info    Price  
0    non-stop         No info   3897.0  
1     2 stops         No info   7662.0  
2     2 stops         No info  13882.0  
3      1 stop         No info   6218.0  
4      1 stop         No info  13302.0

In [23]:
data.drop(columns=['Date_of_Journey'], axis=1,inplace=True)

In [24]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  \
0       IndiGo      24         3     2019  Banglore   New Delhi   
1    Air India       1         5     2019   Kolkata    Banglore   
2  Jet Airways       9         6     2019     Delhi      Cochin   
3       IndiGo      12         5     2019   Kolkata    Banglore   
4       IndiGo       1         3     2019  Banglore   New Delhi   

                   Route Dep_Time  Arrival_Time Duration Total_Stops  \
0              BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1  CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2  DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3        CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4        BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   

  Additional_Info    Price  
0         No info   3897.0  
1         No info   7662.0  
2         No info  13882.0  
3         No info   6218.0  
4         No info  13302.0

## SPLITTING THE ARRIVAL TIME AND REMOVING THE UNECESSARY DATE 

In [25]:
at = []
for time in data['Arrival_Time']:
    time = str(time).split(' ')[0]
    at.append(time)



In [26]:
data.drop(columns=['Arrival_Time'],axis=1, inplace=True)

In [27]:
data.insert(loc=8,column='Arrival_Time', value=at)

In [28]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  \
0       IndiGo      24         3     2019  Banglore   New Delhi   
1    Air India       1         5     2019   Kolkata    Banglore   
2  Jet Airways       9         6     2019     Delhi      Cochin   
3       IndiGo      12         5     2019   Kolkata    Banglore   
4       IndiGo       1         3     2019  Banglore   New Delhi   

                   Route Dep_Time Arrival_Time Duration Total_Stops  \
0              BLR → DEL    22:20        01:10   2h 50m    non-stop   
1  CCU → IXR → BBI → BLR    05:50        13:15   7h 25m     2 stops   
2  DEL → LKO → BOM → COK    09:25        04:25      19h     2 stops   
3        CCU → NAG → BLR    18:05        23:30   5h 25m      1 stop   
4        BLR → NAG → DEL    16:50        21:35   4h 45m      1 stop   

  Additional_Info    Price  
0         No info   3897.0  
1         No info   7662.0  
2         No info  13882.0  
3         No info   6218.0  
4         No info  13302.0

## NOW SPLITTING ARRIVAL AND DEPARTURE TIME INTO HOURS AND MINUTES

In [29]:
arr_hrs = data['Arrival_Time'].str.split(':').str[0].astype(int)
arr_mins = data['Arrival_Time'].str.split(':').str[1].astype(int)
dep_hrs = data['Dep_Time'].str.split(':').str[0].astype(int)
dep_mins = data['Dep_Time'].str.split(':').str[1].astype(int)

In [30]:
data.drop(columns=['Arrival_Time','Dep_Time'], axis=1, inplace=True)

In [31]:
data.insert(loc=7,column='Dep_Hours', value=dep_hrs)
data.insert(loc=8,column='Dep_Mins', value=dep_mins)
data.insert(loc=9,column='Arrival_Hours', value=arr_hrs)
data.insert(loc=10,column='Arrival_Min', value=arr_mins)

In [32]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  \
0       IndiGo      24         3     2019  Banglore   New Delhi   
1    Air India       1         5     2019   Kolkata    Banglore   
2  Jet Airways       9         6     2019     Delhi      Cochin   
3       IndiGo      12         5     2019   Kolkata    Banglore   
4       IndiGo       1         3     2019  Banglore   New Delhi   

                   Route  Dep_Hours  Dep_Mins  Arrival_Hours  Arrival_Min  \
0              BLR → DEL         22        20              1           10   
1  CCU → IXR → BBI → BLR          5        50             13           15   
2  DEL → LKO → BOM → COK          9        25              4           25   
3        CCU → NAG → BLR         18         5             23           30   
4        BLR → NAG → DEL         16        50             21           35   

  Duration Total_Stops Additional_Info    Price  
0   2h 50m    non-stop         No info   3897.0  
1   7h 25m     2 stops         No info   7662.0  
2      19h     2 stops         No info  13882.0  
3   5h 25m      1 stop         No info   6218.0  
4   4h 45m      1 stop         No info  13302.0

In [33]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13353 entries, 0 to 2670
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13353 non-null  object 
 1   JR_Day           13353 non-null  int64  
 2   JR_Month         13353 non-null  int64  
 3   JR_Year          13353 non-null  int64  
 4   Source           13353 non-null  object 
 5   Destination      13353 non-null  object 
 6   Route            13353 non-null  object 
 7   Dep_Hours        13353 non-null  int64  
 8   Dep_Mins         13353 non-null  int64  
 9   Arrival_Hours    13353 non-null  int64  
 10  Arrival_Min      13353 non-null  int64  
 11  Duration         13353 non-null  object 
 12  Total_Stops      13353 non-null  object 
 13  Additional_Info  13353 non-null  object 
 14  Price            10682 non-null  float64
dtypes: float64(1), int64(7), object(7)
memory usage: 1.6+ MB


## CONVERTING THE DATA TYPE OF TOTAL STOPS FROM OBJECT TO INTEGER

In [34]:
data['Total_Stops'] = data['Total_Stops'].map(
    {
        '1 stop'  : 1,
        'non-stop'  : 0,
        '2 stops'  : 2,
        '3 stops'  : 3,
        '4 stops'  : 4
    }
)

In [35]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  \
0       IndiGo      24         3     2019  Banglore   New Delhi   
1    Air India       1         5     2019   Kolkata    Banglore   
2  Jet Airways       9         6     2019     Delhi      Cochin   
3       IndiGo      12         5     2019   Kolkata    Banglore   
4       IndiGo       1         3     2019  Banglore   New Delhi   

                   Route  Dep_Hours  Dep_Mins  Arrival_Hours  Arrival_Min  \
0              BLR → DEL         22        20              1           10   
1  CCU → IXR → BBI → BLR          5        50             13           15   
2  DEL → LKO → BOM → COK          9        25              4           25   
3        CCU → NAG → BLR         18         5             23           30   
4        BLR → NAG → DEL         16        50             21           35   

  Duration  Total_Stops Additional_Info    Price  
0   2h 50m            0         No info   3897.0  
1   7h 25m            2         No info   7662.0  
2      19h            2         No info  13882.0  
3   5h 25m            1         No info   6218.0  
4   4h 45m            1         No info  13302.0

In [36]:
data.drop(columns=['Route'],axis=1,inplace=True)

In [37]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  Dep_Hours  \
0       IndiGo      24         3     2019  Banglore   New Delhi         22   
1    Air India       1         5     2019   Kolkata    Banglore          5   
2  Jet Airways       9         6     2019     Delhi      Cochin          9   
3       IndiGo      12         5     2019   Kolkata    Banglore         18   
4       IndiGo       1         3     2019  Banglore   New Delhi         16   

   Dep_Mins  Arrival_Hours  Arrival_Min Duration  Total_Stops Additional_Info  \
0        20              1           10   2h 50m            0         No info   
1        50             13           15   7h 25m            2         No info   
2        25              4           25      19h            2         No info   
3         5             23           30   5h 25m            1         No info   
4        50             21           35   4h 45m            1         No info   

     Price  
0   3897.0  
1   7662.0  
2  13882.0  
3   6218.0  
4  13302.0

In [38]:
data.shape

(13353, 14)

## CONVERTING THE DURATION INTO MINS

In [39]:
dur = data['Duration']
min_dur = []
for d in dur:
    d = str(d.replace('h','').replace('m', ''))
    d = d.split(' ')

    if len(d) == 1:
        d = int(d[0])
        if d < 60 :
            pass
        else:
            d = d*60
    else:
        d = int(d[0])*60 + int(d[1])
    min_dur.append(d)


In [40]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  Dep_Hours  \
0       IndiGo      24         3     2019  Banglore   New Delhi         22   
1    Air India       1         5     2019   Kolkata    Banglore          5   
2  Jet Airways       9         6     2019     Delhi      Cochin          9   
3       IndiGo      12         5     2019   Kolkata    Banglore         18   
4       IndiGo       1         3     2019  Banglore   New Delhi         16   

   Dep_Mins  Arrival_Hours  Arrival_Min Duration  Total_Stops Additional_Info  \
0        20              1           10   2h 50m            0         No info   
1        50             13           15   7h 25m            2         No info   
2        25              4           25      19h            2         No info   
3         5             23           30   5h 25m            1         No info   
4        50             21           35   4h 45m            1         No info   

     Price  
0   3897.0  
1   7662.0  
2  13882.0  
3   6218.0  
4  13302.0

In [41]:
data.insert(loc=11, column='Duration in Minutes', value=min_dur)


In [42]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  Dep_Hours  \
0       IndiGo      24         3     2019  Banglore   New Delhi         22   
1    Air India       1         5     2019   Kolkata    Banglore          5   
2  Jet Airways       9         6     2019     Delhi      Cochin          9   
3       IndiGo      12         5     2019   Kolkata    Banglore         18   
4       IndiGo       1         3     2019  Banglore   New Delhi         16   

   Dep_Mins  Arrival_Hours  Arrival_Min Duration  Duration in Minutes  \
0        20              1           10   2h 50m                  170   
1        50             13           15   7h 25m                  445   
2        25              4           25      19h                   19   
3         5             23           30   5h 25m                  325   
4        50             21           35   4h 45m                  285   

   Total_Stops Additional_Info    Price  
0            0         No info   3897.0  
1            2         No info   7662.0  
2            2         No info  13882.0  
3            1         No info   6218.0  
4            1         No info  13302.0

In [43]:
data.drop(columns=['Duration'],axis=1,inplace=True)

In [44]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  Dep_Hours  \
0       IndiGo      24         3     2019  Banglore   New Delhi         22   
1    Air India       1         5     2019   Kolkata    Banglore          5   
2  Jet Airways       9         6     2019     Delhi      Cochin          9   
3       IndiGo      12         5     2019   Kolkata    Banglore         18   
4       IndiGo       1         3     2019  Banglore   New Delhi         16   

   Dep_Mins  Arrival_Hours  Arrival_Min  Duration in Minutes  Total_Stops  \
0        20              1           10                  170            0   
1        50             13           15                  445            2   
2        25              4           25                   19            2   
3         5             23           30                  325            1   
4        50             21           35                  285            1   

  Additional_Info    Price  
0         No info   3897.0  
1         No info   7662.0  
2         No info  13882.0  
3         No info   6218.0  
4         No info  13302.0

## LABEL ENCODING THE REMAINING CATEGORICAL DATESETS : SOURCE,DESTINATION , AIRLINES, ADDITIONAL INFO

In [45]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [46]:
data['Airline'] = le.fit_transform(data['Airline'])

In [47]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  Dep_Hours  \
0        3      24         3     2019  Banglore   New Delhi         22   
1        1       1         5     2019   Kolkata    Banglore          5   
2        4       9         6     2019     Delhi      Cochin          9   
3        3      12         5     2019   Kolkata    Banglore         18   
4        3       1         3     2019  Banglore   New Delhi         16   

   Dep_Mins  Arrival_Hours  Arrival_Min  Duration in Minutes  Total_Stops  \
0        20              1           10                  170            0   
1        50             13           15                  445            2   
2        25              4           25                   19            2   
3         5             23           30                  325            1   
4        50             21           35                  285            1   

  Additional_Info    Price  
0         No info   3897.0  
1         No info   7662.0  
2         No info  13882.0  
3         No info   6218.0  
4         No info  13302.0

In [48]:
data.Airline.unique()

array([ 3,  1,  4,  8,  6,  2, 10,  0, 11,  5,  7,  9])

In [49]:
data['Source'] = le.fit_transform(data['Source'])
data['Destination'] = le.fit_transform(data['Destination'])
data['Additional_Info'] = le.fit_transform(data['Additional_Info'])

In [50]:
data.head()

Airline  JR_Day  JR_Month  JR_Year  Source  Destination  Dep_Hours  \
0        3      24         3     2019       0            5         22   
1        1       1         5     2019       3            0          5   
2        4       9         6     2019       2            1          9   
3        3      12         5     2019       3            0         18   
4        3       1         3     2019       0            5         16   

   Dep_Mins  Arrival_Hours  Arrival_Min  Duration in Minutes  Total_Stops  \
0        20              1           10                  170            0   
1        50             13           15                  445            2   
2        25              4           25                   19            2   
3         5             23           30                  325            1   
4        50             21           35                  285            1   

   Additional_Info    Price  
0                8   3897.0  
1                8   7662.0  
2                8  13882.0  
3                8   6218.0  
4                8  13302.0

In [51]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13353 entries, 0 to 2670
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Airline              13353 non-null  int64  
 1   JR_Day               13353 non-null  int64  
 2   JR_Month             13353 non-null  int64  
 3   JR_Year              13353 non-null  int64  
 4   Source               13353 non-null  int64  
 5   Destination          13353 non-null  int64  
 6   Dep_Hours            13353 non-null  int64  
 7   Dep_Mins             13353 non-null  int64  
 8   Arrival_Hours        13353 non-null  int64  
 9   Arrival_Min          13353 non-null  int64  
 10  Duration in Minutes  13353 non-null  int64  
 11  Total_Stops          13353 non-null  int64  
 12  Additional_Info      13353 non-null  int64  
 13  Price                10682 non-null  float64
dtypes: float64(1), int64(13)
memory usage: 1.5 MB


In [52]:
data.isnull().sum()

Airline                   0
JR_Day                    0
JR_Month                  0
JR_Year                   0
Source                    0
Destination               0
Dep_Hours                 0
Dep_Mins                  0
Arrival_Hours             0
Arrival_Min               0
Duration in Minutes       0
Total_Stops               0
Additional_Info           0
Price                  2671
dtype: int64

## SPLITTING DATA INTO TESTING SET AND TRAINING SET

In [53]:
test_data = data[data['Price'].isnull() == True]
train_data = data[data['Price'].isnull() == False]

In [54]:
test_data.head()

Airline  JR_Day  JR_Month  JR_Year  Source  Destination  Dep_Hours  \
0        4       6         6     2019       2            1         17   
1        3      12         5     2019       3            0          6   
2        4      21         5     2019       2            1         19   
3        6      21         5     2019       2            1          8   
4        0      24         6     2019       0            2         23   

   Dep_Mins  Arrival_Hours  Arrival_Min  Duration in Minutes  Total_Stops  \
0        30              4           25                  655            1   
1        20             10           20                    4            1   
2        15             19            0                 1425            1   
3         0             21            0                   13            1   
4        55              2           45                  170            0   

   Additional_Info  Price  
0                8    NaN  
1                8    NaN  
2                5    NaN  
3                8    NaN  
4                8    NaN

In [55]:
train_data.head()

Airline  JR_Day  JR_Month  JR_Year  Source  Destination  Dep_Hours  \
0        3      24         3     2019       0            5         22   
1        1       1         5     2019       3            0          5   
2        4       9         6     2019       2            1          9   
3        3      12         5     2019       3            0         18   
4        3       1         3     2019       0            5         16   

   Dep_Mins  Arrival_Hours  Arrival_Min  Duration in Minutes  Total_Stops  \
0        20              1           10                  170            0   
1        50             13           15                  445            2   
2        25              4           25                   19            2   
3         5             23           30                  325            1   
4        50             21           35                  285            1   

   Additional_Info    Price  
0                8   3897.0  
1                8   7662.0  
2                8  13882.0  
3                8   6218.0  
4                8  13302.0

In [97]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

rfr = RandomForestRegressor()
X = train_data.iloc[:,:-1]
Y = train_data.iloc[:,-1]
X_train , X_test, y_train, y_test = train_test_split(X,Y,test_size=0.15, random_state=10)
rfr.fit(X_train, y_train)

RandomForestRegressor()

In [98]:
print(X_train.shape,y_train.shape)

(9079, 13) (9079,)


In [99]:
y_pred = rfr.predict(X_test)
print("MSE:", round(mean_squared_error(y_pred,y_test),3))
print("R2:", round(r2_score(y_pred, y_test),2))
print("--------------")


for i in y_pred:
    print(i)



MSE: 2986410.58
R2: 0.84
--------------
7962.12
13989.59
10501.25
7009.36
11547.99
10477.81
12929.636166666667
3090.88
4713.06
3272.19
6893.61
5666.698333333334
13376.0
3871.96
5051.75
12317.44
13063.03
4407.67
10173.04
6378.47
5398.37
11289.28
14201.879095238093
8037.36
6893.7
2640.62
7593.74
13052.82
7472.13
8156.63
4829.82
10178.82
8522.6
4369.07
9552.7
7128.8825
9153.53
5475.21
12504.96
16931.6
8700.66
7520.36
10885.28
5775.9
13525.22
3588.43
9723.47
5643.74
10653.96
7147.83
6248.06
13841.95
12398.28
6172.03
5516.325
11356.5075
4003.08
10752.9
4946.25
3943.0
10262.0
8543.96
3840.74
8042.25
7627.15
5310.04
7229.0
10265.15
12721.79
14754.9
13005.97
14859.65
11130.185
10899.99
4177.19
14269.02
4377.74
3601.4
12031.44
13156.464761904763
15709.43
14099.61
6760.26
13947.31
13666.57
13139.503333333332
4307.58
6791.37
8766.47
12748.389333333336
10335.696499999998
2961.48
3306.16
3860.4
12037.155
5084.86
16320.5
2758.21
11090.16
2111.99
10551.96
3906.8
13914.87
11680.36
5561.88
3600.9
5892.

In [100]:
for i in y_test.values:
    print(i)

6224.0
14151.0
10539.0
7934.0
16754.0
11447.0
11522.0
3100.0
4030.0
3100.0
9125.0
7608.0
13376.0
3873.0
5823.0
12681.0
14151.0
7417.0
9899.0
6327.0
6069.0
12534.0
13014.0
7832.0
6961.0
2647.0
7952.0
16932.0
9165.0
7888.0
4823.0
9899.0
7927.0
4284.0
9314.0
6138.0
9134.0
5281.0
11038.0
17024.0
9187.0
5678.0
11087.0
5228.0
14151.0
3540.0
9663.0
5678.0
7664.0
8075.0
6478.0
13626.0
11642.0
6117.0
6205.0
11150.0
3943.0
10703.0
4804.0
3943.0
10262.0
8452.0
3841.0
8016.0
7670.0
4823.0
7229.0
10262.0
18228.0
16079.0
13067.0
16372.0
11507.0
10919.0
4174.0
13017.0
3359.0
3597.0
10703.0
13014.0
15147.0
13941.0
7452.0
11600.0
13591.0
13014.0
4488.0
6478.0
9526.0
12760.0
9718.0
2754.0
3175.0
3807.0
9920.0
4990.0
17024.0
2754.0
11087.0
2175.0
10262.0
3419.0
14571.0
9586.0
3948.0
3597.0
5298.0
3100.0
13626.0
14714.0
10844.0
9585.0
11752.0
6300.0
7308.0
9748.0
8016.0
12692.0
10678.0
10262.0
6112.0
10844.0
8366.0
22270.0
6253.0
3175.0
14714.0
7198.0
5192.0
12723.0
7229.0
15419.0
10262.0
10262.0
10844.0


In [105]:
X

Airline  JR_Day  JR_Month  JR_Year  Source  Destination  Dep_Hours  \
0            3      24         3     2019       0            5         22   
1            1       1         5     2019       3            0          5   
2            4       9         6     2019       2            1          9   
3            3      12         5     2019       3            0         18   
4            3       1         3     2019       0            5         16   
...        ...     ...       ...      ...     ...          ...        ...   
10678        0       9         4     2019       3            0         19   
10679        1      27         4     2019       3            0         20   
10680        4      27         4     2019       0            2          8   
10681       10       1         3     2019       0            5         11   
10682        1       9         5     2019       2            1         10   

       Dep_Mins  Arrival_Hours  Arrival_Min  Duration in Minutes  Total_Stops  \
0            20              1           10                  170            0   
1            50             13           15                  445            2   
2            25              4           25                   19            2   
3             5             23           30                  325            1   
4            50             21           35                  285            1   
...         ...            ...          ...                  ...          ...   
10678        55             22           25                  150            0   
10679        45             23           20                  155            0   
10680        20             11           20                    3            0   
10681        30             14           10                  160            0   
10682        55             19           15                  500            2   

       Additional_Info  
0                    8  
1                    8  
2                    8  
3                    8  
4                    8  
...                ...  
10678                8  
10679                8  
10680                8  
10681                8  
10682                8  

[10682 rows x 13 columns]

In [115]:
x1 = X.iloc[:531,:]
x2 = X.iloc[531:,:]

In [116]:
y1 = Y.iloc[:531]
y2 = Y.iloc[531:]

#### Training Part 1

In [135]:
rr = RandomForestRegressor()
xtr,xt,ytr,yt = train_test_split(x1,y1,test_size=0.3,random_state=10)

In [136]:
rr.fit(xtr,ytr)

RandomForestRegressor()

In [137]:
ypr1 = rr.predict(xt)

In [138]:
for i in ypr1:
    print(i)

11065.48
13256.71
13903.74
10105.15
10563.02
4964.47
10482.97
4935.59
8971.99
4877.2
11707.34
9546.04
3962.39
4928.85
11186.85
15155.0
10741.16
6246.87
14563.75
9878.54
5392.71
6206.93
5565.3
5312.2
15013.72
7039.14
11905.71
4449.45
6696.41
10625.9
8136.1
11057.81
4605.32
11151.94
22976.39
6448.58
6486.09
16599.08
7718.99
4834.05
4277.93
11043.97
3617.77
11177.2
16401.29
3822.27
5608.31
12117.49
10802.87
14818.51
6763.17
11468.15
14676.43
6623.31
10586.61
11186.89
12762.74
14026.71
6705.41
14773.66
13125.43
15794.5
5013.38
7469.88
9082.45
5684.05
10456.3
13211.65
5476.72
4983.57
3117.6
13251.0
6046.8
12842.25
7835.32
4464.26
7416.26
3604.69
14058.39
6897.13
10830.86
4606.99
14347.19
4711.58
11196.26
7779.89
14476.56
5368.64
17087.37
11903.24
3600.46
14149.13
6529.4
9661.4
15879.69
4749.97
6497.33
5207.81
9510.8
11588.07
5538.43
6282.77
8396.58
13422.03
10669.51
4391.53
8500.72
9928.34
11465.32
7925.21
4604.01
13411.93
13906.08
10167.55
11262.49
11728.66
11850.17
3731.13
11271.66
4597.8

In [134]:
for i in yt.values:
    print(i)

15318.0
9483.0
14388.0
7757.0
8586.0
4239.0
10262.0
4667.0
7064.0
3898.0
12898.0
4487.0
3311.0
4174.0
7574.0
8738.0
6795.0
5894.0
8899.0
9899.0
3971.0
4148.0
5264.0
7229.0
18799.0
5162.0
10861.0
3383.0
6442.0
6601.0
3543.0
11087.0
5216.0
10262.0
22270.0
4804.0
7754.0
17024.0
7295.0
5769.0
4077.0
10262.0
2754.0
13381.0
22096.0
3100.0
3597.0
10197.0
10844.0
16079.0
5583.0
11507.0
14641.0
6528.0
9663.0
12639.0
8452.0
13555.0
5618.0
16079.0
13044.0
17024.0
3943.0
8855.0
8991.0
5201.0
12524.0
11507.0
8224.0
3859.0
8040.0
13044.0
7229.0
12242.0
5177.0
3943.0
6197.0
4995.0
15554.0
7229.0
10791.0
3398.0
14714.0
4462.0
9270.0
5281.0
17057.0
4878.0
7378.0
12303.0
1965.0
14781.0
5636.0
6842.0
17024.0
5678.0
6412.0
4823.0
9345.0
13465.0
5830.0
4823.0
8016.0
13731.0
6795.0
3858.0
8366.0
9663.0
11087.0
5013.0
4777.0
15129.0
12677.0
10262.0
12898.0
11087.0
12373.0
5678.0
13591.0
4804.0
7662.0
10844.0
5117.0
8907.0
10262.0
7229.0
3858.0
10262.0
11432.0
5228.0
7202.0
12373.0
11596.0
